### PR TITLE
add AddStep() to WaterfallDialog and update unit tests

### PR DESCRIPTION
**Pull request by tomlm**

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 1]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
 
-        private WaterfallStep[] _steps;
+        private List<WaterfallStep> _steps;
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WaterfallDialog"/> class.
+        /// </summary>
+        /// <param name="dialogId">dialog id</param>
+        /// <param name="steps">optional steps to be defined by caller</param>
         public WaterfallDialog(string dialogId, WaterfallStep[] steps = null)
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: its kind of awkward to receive an array and convert it to list. Do people create the array just for this class? if yes, maybe receive a list? This is fine too, just thinking what's best.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 2]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
         private const string StepIndex = "stepIndex";
         private const string PersistedValues = "values";
 
-        private WaterfallStep[] _steps;
+        private List<WaterfallStep> _steps;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: readonly. As far as I can tell, this is only ever set in the constructor.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 3]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WaterfallDialog"/> class.
+        /// </summary>
+        /// <param name="dialogId">dialog id</param>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: There is room to raise the bar on this particular comment. :)

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 4]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
+        /// <param name="steps">optional steps to be defined by caller</param>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: Please make this a full sentence, with correct casing.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs (hunk 1 above)
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: +1 to what Carlos is saying. I would either:
1. Have this class take a list. 
2. Use the Params to devs can just pass in "Step1, Step2, Step3". 

I seldom see C# Developers using arrays...

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 5]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
+            {
+                _steps = new List<WaterfallStep>();
+            }
+        }
+
+        /// <summary>
+        /// Add a new step to the waterfall.
+        /// </summary>
+        /// <param name="step">step to add</param>
+        /// <returns>waterfall dialog for fluent calls to .AddStep()</returns>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: Grammar. Please make this a full sentence.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 6]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
+        public WaterfallDialog AddStep(WaterfallStep step)
+        {
+            this._steps.Add(step);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: no need for "this.".

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs [hunk 7]
```diff
@@ -20,12 +20,35 @@ public class WaterfallDialog : Dialog
...
+            }
+        }
+
+        /// <summary>
+        /// Add a new step to the waterfall.
+        /// </summary>
+        /// <param name="step">step to add</param>
+        /// <returns>waterfall dialog for fluent calls to .AddStep()</returns>
+        public WaterfallDialog AddStep(WaterfallStep step)
+        {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: null check + throw on step.

Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: ... and please don't forget the Unit test. :)

File: tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs [hunk 8]
```diff
@@ -277,4 +309,32 @@ private static async Task<DialogTurnResult> Waterfall5_Step2(DialogContext dc, W
...
+            await dc.Context.SendActivityAsync("step2");
+            return Dialog.EndOfTurn;
+        }
+        private static async Task<DialogTurnResult> Waterfall2_Step3(DialogContext dc, WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            await dc.Context.SendActivityAsync("step3");
+            return Dialog.EndOfTurn; 
+        }
+    }
+
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs]: Extra blank line.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs (hunk 1 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: good point, I changed to IEnumerable<> that covers [], list, set, etc.

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs (hunk 3 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: That is what stylecop insists it should be :P''''

File: libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs (hunk 6 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/WaterfallDialog.cs]: except for the enhanced readability.

**Pull request merged**